### DISCUSSION TITLE
Offset footprint with angle in mind

### DISCUSSION THREAD

**Discussion by ReFil** [INTI-CMNB/KiBot#591] [Q&A] [answered]
I'm trying to automate correcting the origin for some Choc hotswap sockets, however it seems like the `rot_footprint` filter doesn't offset with respect to the current angle of the component, it just applies a fixed x and Y offset, which leads to inconsistent positioning depending on the angle of the component, and doesn't work at all on a board with the same footprint at multiple angles. Is there a flag i'm missing to have it respect the current angle of the component?

**Comment by set-soft** [comment]
Please create a small example showing your issue. I.e. containing 2 components with different rotation. Include the schematic, PCB and KiBot config. 

The code takes the rotation:

```python
    def apply_offset_value(self, comp, angle, pos_offset_x, pos_offset_y):
        if angle:
            rotation = radians(angle)
            rsin = sin(rotation)
            rcos = cos(rotation)
            comp.pos_offset_x = pos_offset_x * rcos - pos_offset_y * rsin
            comp.pos_offset_y = pos_offset_x * rsin + pos_offset_y * rcos
            logger.debugl(2, f'- rotating offset {angle} degrees: {comp.pos_offset_x}, {comp.pos_offset_y} mm')
        else:
            comp.pos_offset_x = pos_offset_x
            comp.pos_offset_y = pos_offset_y
        # The signs here matches matthewlai/JLCKicadTools offsets because the database comes from this plugin
        comp.pos_offset_x = -GS.from_mm(comp.pos_offset_x)
        comp.pos_offset_y = GS.from_mm(comp.pos_offset_y)
        logger.debugl(2, f'- final offset {comp.pos_offset_x}, {comp.pos_offset_y} KiCad IUs')
```

But it has too many options to provide compatibility with different tools. So the combination of options is really important.

**Reply by ReFil** [reply]
Hi, please find it here. Here's a picture showing what happened with JLCPCB cpl too to demonstrate the issue
![image](https://github.com/INTI-CMNB/KiBot/assets/31960031/f623ec04-32c2-4f90-a4cf-8049447f287d)
[angle-problem-demo.zip](https://github.com/INTI-CMNB/KiBot/files/14697198/angle-problem-demo.zip)

**Reply by set-soft** [reply]
If you run KiBot using `-L logfile` you'll get a detailed log, including:

```
DEBUG:- rotating CH2 from 160.0 to 200.0 (kibot - fil_rot_footprint.py:227)
DEBUG:- matched re.compile('choc_v1_HS') on field footprint with offset -2.5, -4.85 mm (kibot - fil_rot_footprint.py:299)
DEBUG:- rotating offset 200.0 degrees: 0.6904338568352784, 5.412559569125827 mm (kibot - fil_rot_footprint.py:263)
```

Here you can see the offset is rotated by 200º. I think the issue is with your custom rotation filter:

```
  - name: fix_jlc_rotation
    comment: 'Adjust rotation'
    type: rot_footprint
    negative_bottom: false
    invert_bottom: true
    rotations:
      - ["^SOD-323*", 180.0]
      - ["^XKB-USB-C-16P", 180]
    rotations_and_offsets:
      - regex: 'choc_v1_HS'
        offset_x: -2.5
        offset_y: -4.85
      - regex: 'MX_HS'
        offset_x: 0
        offset_y: -3.81
```

This doesn't match the current internal default filter. The default filter doesn't invert the bottom, instead it mirrors the bottom using  *mirror_bottom*.

This was changed in the last months, from what I found is not the first time JLCPCB changes how to compute the rotations for bottom components. As they use EasyEDA they sync the changes with it, but these changes breaks KiCad plug-ins. 

I adjusted KiBot to match the last change, this is why the internal filter is updated.

**Reply by set-soft** [reply]
P.S. Looking at the docs I see it said *mirror_bottom: false*  instead of *mirror_bottom: true*. Now fixed.

**Reply by ReFil** [reply]
Ok, so I need to remove the ~~`mirror_bottom`~~ `invert_bottom` flag?

**Reply by set-soft** [reply]
Yes, and also add `mirror_bottom: true`

**Reply by set-soft** [reply]
Is currently documented: [Docs](https://kibot.readthedocs.io/en/latest/notes_position.html)

**Reply by ReFil** [reply]
Unfortunately that hasn't solved it, I had to change the +ve offsets to -ve in the yaml file so the unrotated component was in the correct location but the rotated one is in the same position it was before. By my maths it should be moved 3.8 and 3.72 mm  in x and y respectively but instead it is being moved 0.69 and 5.412 as it is currently according to the logs. did i make a different mistake in the file
[angle-problem-demo.zip](https://github.com/INTI-CMNB/KiBot/files/14724020/angle-problem-demo.zip)

**Reply by set-soft** [reply]
Which math are you using?

**Reply by ReFil** [reply]
My apologies for the crude diagram, here's how I calculated my x and y offsets. 
The absolute distance from the origin to the offset origin remains the same irrelevant of the rotation of the component (5.456mm), and the angle is referenced to the vertical axis (27.27 degrees between the hypotenuse and downward y axis). 
I applied the component's rotation in polar coordinate space (so 47.27 degrees between hypotenuse and downwards y axis), then converted back to cartesian. 
I made a slight mistake in the maths in my previous message, it should be 3.702 y and 4.008 x offset using this method, which visually looks to be more correct. I think the current method applies the current rotation in the other direction as it is 7.27 degrees between hypotenuse and downwards y axis but length is correct.

![20240325_170612.jpg](https://github.com/INTI-CMNB/KiBot/assets/31960031/8a8bed7c-47b6-4471-addc-891e151f1200)

**Reply by set-soft** [reply, marked as answer]
This is the computation KiBot is doing, just with the reverse angle sign. The problem is that this must work for other components. Not just this one. I'm trying with C2935951 which needs both: rotation and offset, and I can figure out how to handle it yet.
Note that you can use schematic fields to control the offset and rotation of individual components to handle cases where KiBot can solve the problem. In this case I recommend using `bennymeg_mode: false`.

**Reply by ReFil** [reply]
Ok no worries, thank you for all your assistance

**Reply by set-soft** [reply]
Ok, I think I found the problem. For the bottom we need to change the angle sign and also the X offset sign. This solves the problem.

I added a test case to the repo [here](https://github.com/INTI-CMNB/KiBot/tree/dev/experiments/JLC/Test_Case). This includes rotations, offsets and rotations+offsets for the top and bottom.

With the last fix (found in *dev* branch and images) the test case seems to work really well when uploaded to JLCPCB.

Top side:

![image](https://github.com/INTI-CMNB/KiBot/assets/22823627/672831b8-34b5-401c-bf71-1a51165c3a78)

Bottom side:

![image](https://github.com/INTI-CMNB/KiBot/assets/22823627/6abbe163-eecf-4c74-8359-4258c46dc882)

As you can see I included the component you used. Note that I used the following filter:

```yaml
  - name: rotations
    comment: 'Test case'
    type: rot_footprint
    negative_bottom: false
    mirror_bottom: true
    rotations_and_offsets:
       - regex: 'choc_v1_HS'
         offset_x: -2.5
         offset_y: 4.85
```
